File: TypeScript/src/Entities/EntityFactory.ts
import { Entity } from "./Entity";
import { Cat } from "./Cat";
import { Dog } from "./Dog";
import { CatAdapter } from "./CatAdapter";
import { Person } from "./Person";

export class EntityFactory {
    private readonly catNames: string[] = ["Sally", "Kitty", "Raider", "Rosie"];
    private readonly dogNames: string[] = ["Rover", "Rex", "Spot", "Fido"];
    private readonly peopleNames: string[] = ["John", "Adam", "Emily", "Anna"];

    public createCat(): Entity {
        return new Cat(this.getRandomName(this.catNames));
    }

    public createDog(): Entity {
        return new Dog(this.getRandomName(this.dogNames));
    }

    public createDisguisedDog(): Entity {
        return new CatAdapter(new Dog(this.getRandomName(this.dogNames)));
    }

    public createPerson(): Person {
        return new Person(this.getRandomName(this.peopleNames));
    }

    private getRandomName(options: string[]): string {
        const randomIndex = Math.floor(Math.random() * options.length);
        return options[randomIndex];
    }
}

File: Java/src/java/Entities/EntityFactory.java
public class EntityFactory {

    private List<String> catNames = List.of("Sally", "Kitty", "Raider", "Rosie");
    private List<String> dogNames = List.of("Rover", "Rex", "Spot", "Fido");
    private List<String> peopleNames = List.of("John", "Adam", "Emily", "Anna");

    public Entity createCat(String name) {
        return new Cat(getRandomName(catNames));
    }

    public Entity createDog(String name) {
        return new Dog(getRandomName(dogNames));
    }

    public Entity createDisguisedDog() {
        return new CatAdapter(new Dog(getRandomName(dogNames)));
    }

    public CanAdmire createPerson() {
        return new Person(getRandomName(peopleNames));
    }

    private getRandomName(List<String> options) {
        return options.get(new Random().nextInt(options.size()))
    }
}